File: tasks/money-rollup/go/src/fx.go
package main

import "errors"

func edgeKey(source string, target string) string {
	return source + " " + target
}

func buildGraph(currencies map[string]int, rates []rate) (map[string]float64, error) {
	edges := map[string]float64{}
	for _, edge := range rates {
		value, err := parseRate(edge.Rate)
		if err != nil {
			return nil, err
		}
		edges[edgeKey(edge.From, edge.To)] = value
	}
	return edges, nil
}

func factor(edges map[string]float64, source string, target string) (float64, error) {
	if source == target {
		return 1, nil
	}
	direct, present := edges[edgeKey(source, target)]
	if !present {
		return 0, errors.New("no conversion rate")
	}
	return direct, nil
}

File: tasks/money-rollup/python-typed/solution/fx.py
"""Directed currency rate graph and exact conversion factors."""

import os
from fractions import Fraction

from money import parse_rate

Graph = dict[tuple[str, str], Fraction]


def build_graph(currencies: dict[str, int], rates: list[object]) -> Graph:
    edges: Graph = {}
    for rate in rates:
        if not isinstance(rate, dict) or set(rate) != {"from", "to", "rate"}:
            raise ValueError("malformed rate")
        source, target = rate["from"], rate["to"]
        if source not in currencies or target not in currencies:
            raise ValueError("unknown rate currency")
        if source == target:
            raise ValueError("self rate")
        if (source, target) in edges:
            raise ValueError("duplicate rate")
        edges[(source, target)] = parse_rate(rate["rate"])
    return edges


def factor(edges: Graph, source: str, target: str) -> Fraction:
    """Exact product along the unique shortest directed path."""
    if source == target:
        return Fraction(1)
    if os.environ.get("LAB_SABOTAGE") == "direct-rate-only":
        if (source, target) not in edges:
            raise ValueError("no conversion path")
        return edges[(source, target)]
    reached: dict[str, tuple[Fraction, int]] = {source: (Fraction(1), 1)}
    frontier = [source]
    while frontier:
        following: dict[str, tuple[Fraction, int]] = {}
        for node in frontier:
            base, count = reached[node]
            for (start, end), rate in edges.items():
                if start != node or end in reached:
                    continue
                seen = following.get(end)
                if seen is None:
                    following[end] = (base * rate, count)
                else:
                    following[end] = (seen[0], seen[1] + count)
        if not following:
            break
        reached.update(following)
        if target in reached:
            value, count = reached[target]
            if count != 1:
                raise ValueError("ambiguous conversion path")
            return value
        frontier = list(following)
    raise ValueError("no conversion path")
